### ISSUE TITLE
[RFC] Support for training and evaluation on multiple datasets in `SFTTrainer`

### ISSUE THREAD

**Issue by SumanthRH** [NovaSky-AI/SkyRL#1875]
## Summary

Add native `SFTTrainer` support for training on a weighted mixture of multiple
datasets and evaluating on multiple named datasets. This should be a thin layer on top of the existing `StatefulDataLoader` and still allow for custom samplers on top. Evaluation metrics should be logged per-evaluation dataset.

The intended user experience is:

```bash
bash examples/train/sft/run_sft_megatron.sh \
  train_datasets="['allenai/tulu-3-sft-mixture','yahma/alpaca-cleaned','openchat/openchat_sharegpt4_dataset']" \
  train_dataset_splits="['train[:50000]','train[:10000]','train[:10000]']" \
  train_dataset_weights="[0.8,0.1,0.1]" \
  eval_datasets="['allenai/tulu-3-sft-mixture','yahma/alpaca-cleaned']" \
  eval_dataset_splits="['train[-500:]','train[200:700]']" \
  eval_dataset_names="['tulu3','alpaca']"
```

## Motivation

Often, researchers are experimenting with multiple datasets for SFT. For training, it is common to iterate on different data mixtures for SFT. Further, evaluation is typically performed on multiple datasets, often not only on the datasets present in the training mixture but additional datasets to evaluate model capabilities on other axes, track any catastrophic forgetting, etc.

This RFC details a simple extension to the `SFTTrainer` to support multiple datasets for training and evaluation in SkyRL.

## Goals

- Support multiple training datasets with source-level sampling weights.
- Keep weighted sampling independent of source dataset size.
- Support multiple eval datasets with metrics logged separately per dataset.
- Support custom samplers with multiple training datasets: For example, users can change the dataset mixing weights after training progresses. 
- All datasets should have the same relevant columns for training - `messages_key`, `tools_key` and `system_key`, but can have additional unique columns beyond these.

## Non-goals

- Streaming dataset loading. This RFC keeps the current in-memory tokenized dataset model.
- Dataset-specific preprocessing or tokenization settings.
- Native dynamic mixing schedules. Users can express these through `sampler=custom`.
- Mixing datasets with different modalities or different schemas. All datasets should support the same `messages_key`, `tools_key` and `system_key`. Further, they should all be text-only or with images but not both.

## Proposed config

Add list-based train and eval dataset fields:

```python
class SFTConfig:
    # Training data
    train_datasets: Optional[List[str]] = None
    train_dataset_splits: Optional[List[str]] = None
    train_dataset_weights: Optional[List[float]] = None

    # Evaluation data
    eval_datasets: Optional[List[str]] = None
    eval_dataset_splits: Optional[List[str]] = None
	# optional, shorthand names used only for logging
    eval_dataset_names: Optional[List[str]] = None
```

1. Compatibility:

- `dataset_name` and `dataset_split` are deprecated in favour of `train_datasets` and `train_dataset_splits`. This is similarly true for `eval_datasets` and `eval_dataset_name`.
- SkyRL will still support user-provided overrides with the old fields, but internally translate them to the new list format. Add a deprecation warning in such cases. 

2. What does `train_dataset_weights` mean?
- Ratios of samples from each dataset within a batch (approximately), unrelated to the lengths of the dataset.
- Default ratio is `1/N` where `N` is the number of training datasets -> equal mixing from all the datasets.

## Training design

- Load and tokenize each training dataset independently using the existing
`_load_and_tokenize(dataset_name, dataset_split)` path. This preserves current
tokenization behavior and cache keys. After tokenization, concatenate the lists. Keep track of the lengths of each dataset.
- When there is a single dataset specified, keep the behaviour unchanged: Simply use basic shuffling or user provided sampler for dataloading.
- With multiple datasets: load into a single dataloader as before, but with a [`DataMixingSampler`](https://github.com/NovaSky-AI/SkyRL/blob/1ab51f1b965c0a3a6fa7dfd4e965758525bfd89b/examples/train/sft/data_mixing_sampler.py#L37) by default. Configure the lengths according to the individual dataset sizes and weights according to user specified `train_dataset_weights`.
	- Allow users to provide a custom sampler override. This can choose to respect the ratios or not
	- Ensure that the weighted mixing in `DataMixingSampler` resamples a _fresh_ list of indices every epoch. Currently, sampler's indices list `_plan` is instantiated once and re-used. Update the state of the sampler to include the `rng` state for resumption.
- Checkpointing and resume has no additional considerations: We built on top of the existing dataloading save/resume logic.

## Evaluation design

- Support evaluation on multiple datasets separately. Build one evaluation dataloader per evaluation dataset.
- Run evaluation separately and log metrics individually, grouped under `eval/{eval_dataset_name}`. This nesting will be used even if only one dataset is provided so that users can easily compare WandB metrics across runs with and without dataset mixing.

## Implementation Plan
1. Add support for the new configuration + validation for translating deprecated params
2. Add support for multiple training datasets in the dataset preprocessing stage
3. Move the `DataMixingSampler` into a native sampler in SkyRL and use that in the case with multiple datasets
4. Add support for multiple evaluation datasets as specified above
5. Update documentation with an example for training with multiple datasets and evaluation on multiple datasets. Add an example for user-provided custom sampler on top. 

## Test Plan

The tests can be CPU only tests:
1. Test with user-specified `train_dataset_weights` and `train_datasets`: ensure that representation in a batch is close to the weights specified. 
2. Test with default `train_dataset_weights` and user-specified `train_datasets`: equal mixing
3. Test with user-specified `train_dataset_weights` and `train_datasets` with varying dataset lengths: Mixing ratios are independent of length. Representation in a batch should be roughly unchanged.

**Comment by SumanthRH** [comment]
Assignee: @avigyabb
